### PR TITLE
Stored procedure call returns wrong BigDecimal scale. 

### PR DESCRIPTION
**Pull request by Ananya2**
https://github.com/microsoft/mssql-jdbc/issues/2534 - Set scale value dynamically for DECIMAL output parameters in registerOutParameter method.

This change retrieves the scale from the ParameterMetaData and applies it to the DECIMAL output parameter. This provides a more flexible approach for handling DECIMAL data types. 

As an alternative, users can directly call registerOutParameter(index, sqlType, scale) to specify the scale explicitly.

### CODE REVIEW

**Review by machavan — COMMENTED** [review]
Let us also add a test that resembles the two customer reported scenarios

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerCallableStatement.java [hunk 1]
```diff
@@ -150,6 +151,16 @@ public void registerOutParameter(int index, int sqlType) throws SQLServerExcepti
...
                 param.setOutScale(7);
                 break;
+            case java.sql.Types.DECIMAL:
+                ParameterMetaData parameterMetaData = this.getParameterMetaData();
+                if (parameterMetaData != null) {
+                    try {
+                        // Fetch scale from metadata for DECIMAL type
+                        int scale = parameterMetaData.getScale(index);
+                        param.setOutScale(scale);
+                    } catch (SQLException e) {}
```
Comment (divang) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerCallableStatement.java:161]: Please add the logging for the exception to avoid silent failures.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerCallableStatement.java [hunk 2]
```diff
@@ -150,6 +151,18 @@ public void registerOutParameter(int index, int sqlType) throws SQLServerExcepti
...
                 break;
+            case java.sql.Types.DECIMAL:
+                ParameterMetaData parameterMetaData = this.getParameterMetaData();
+                if (parameterMetaData != null) {
+                    try {
+                        // Fetch scale from metadata for DECIMAL type
+                        int scale = parameterMetaData.getScale(index);
+                        param.setOutScale(scale);
+                    } catch (SQLException e) {
+                       loggerExternal.warning("Failed to fetch scale for DECIMAL type parameter at index " + index + ": " + e.getMessage());
```
Comment (divang) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerCallableStatement.java]: The SQLServerException needs to be explicitly thrown instead of being suppressed.

File: src/test/java/com/microsoft/sqlserver/jdbc/unit/statement/StatementTest.java [hunk 3]
```diff
@@ -1195,15 +1196,66 @@ public void testJdbc41CallableStatementMethods() throws Exception {
...
+        @Test
+        public void testBigDecimalPrecision() throws SQLException {
+            try (Connection connection = getConnection()) {
+                String createProceduresSQL = "create procedure test_bigdecimal_3\n" +
+                "    @big_decimal_type      decimal(15, 3),\n" +
+                "    @big_decimal_type_o    decimal(15, 3) output\n" +
+                "as\n" +
+                "begin\n" +
+                "    set @big_decimal_type_o = @big_decimal_type;\n" +
+                "end;\n";
```
Comment (tkyc) [inline src/test/java/com/microsoft/sqlserver/jdbc/unit/statement/StatementTest.java:1220]: We should move the create and drop of the procedure to the `init()` and `terminate()`. If the test fails for some reason mid run, then the drop call doesn't get executed, and then next run the create fails since the proc already exists. And, the procedure name should be prefixed/suffixed with a unique identifier eg. using `AbstractSQLGenerator.escapeIdentifier`. The create proc might fail if the proc name isn't unique, if multiple runs of this test is running against the same server.

**Comment by tkyc** [comment]
Looks like test `testJdbc41CallableStatementMethods` is failing.

File: src/test/java/com/microsoft/sqlserver/jdbc/unit/statement/StatementTest.java (hunk 3 above)
Comment (lilgreenbird) [reply src/test/java/com/microsoft/sqlserver/jdbc/unit/statement/StatementTest.java:1220]: also, pls use code formatter this doesn't look formatted

**Comment by tkyc** [comment]
pinged you about an internal test fail

**Review by lilgreenbird — CHANGES_REQUESTED** [review]
this failed tests in internal test lab

File: src/test/java/com/microsoft/sqlserver/jdbc/unit/statement/StatementTest.java (hunk 3 above)
Comment (Ananya2) [reply src/test/java/com/microsoft/sqlserver/jdbc/unit/statement/StatementTest.java]: Updated the test case, please review.

File: src/test/java/com/microsoft/sqlserver/jdbc/unit/statement/BigDecimalPrecisionTest.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+            try (Statement stmt = connection.createStatement()) {
+                stmt.execute(createProcedureSQL1);
+                stmt.execute(createProcedureSQL2);
+            }
+        }
+    }
+
+    @AfterEach
+    public void terminate() throws SQLException {
```
Comment (machavan) [inline src/test/java/com/microsoft/sqlserver/jdbc/unit/statement/BigDecimalPrecisionTest.java:53]: Let us use this approach for table creation and dropping in init and terminate methods as it provides idempotency on failures and repeated runs


        @BeforeEach
        public void init() throws Exception {
            try (Connection con = getConnection()) {
                con.setAutoCommit(false);
                try (Statement stmt = con.createStatement()) {
                    TestUtils.dropTableIfExists(tableName, stmt);
                    stmt.executeUpdate(
                            "CREATE TABLE " + tableName + " (col1 INT, col2 VARCHAR(" + TEST_STRING.length() + "))");
                }
                con.commit();
            }
        }



       @AfterEach
        public void terminate() throws Exception {
            try (Connection con = getConnection(); Statement stmt = con.createStatement()) {
                try {
                    TestUtils.dropTableIfExists(tableName, stmt);
                } catch (SQLException e) {
                    fail(TestResource.getResource("R_unexpectedException") + e.getMessage());
                }
            }
        }

**Comment by Ramudaykumar** [comment]
There should be significant comments in code, so that a new person can understand the code.

**Comment by Ananya2** [comment]
> There should be significant comments in code, so that a new person can understand the code.
Noted.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerCallableStatement.java [hunk 5]
```diff
@@ -150,6 +151,21 @@ public void registerOutParameter(int index, int sqlType) throws SQLServerExcepti
...
+                // The scale for the DECIMAL type is fetched from the ParameterMetaData.
+                // This provides flexibility to automatically apply the correct scale as per the database metadata.
+                ParameterMetaData parameterMetaData = this.getParameterMetaData();
+                if (parameterMetaData != null) {
+                    try {
+                        // Fetch scale from metadata for DECIMAL type
+                        int scale = parameterMetaData.getScale(index);
+                        param.setOutScale(scale);
+                    } catch (SQLException e) {
+                        loggerExternal.warning("Failed to fetch scale for DECIMAL type parameter at index " + index + ": " + e.getMessage());
```
Comment (lilgreenbird) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerCallableStatement.java:165]: should this fail? according to the spec this would only throw an error if a database access error occurs, seem quite a serious error

**Review by lilgreenbird — CHANGES_REQUESTED** [review]
doesn't look like the formatter was run on BigDecimalPrecisionTest.java, pls make sure to run formatter on all files

**Review by lilgreenbird — COMMENTED** [review]
there are already BigDecimal tests in StatementTest. Either just add the new tests here or, move all the BigDecimal test to this new file

File: src/test/java/com/microsoft/sqlserver/jdbc/unit/statement/BigDecimalPrecisionTest.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+        try (Connection connection = getConnection()) {
+            String dropProcedureSQL = "DROP PROCEDURE IF EXISTS " + procName1 + ", " + procName2;
+            try (Statement stmt = connection.createStatement()) {
+                stmt.execute(dropProcedureSQL);
+            }
+            
+            String createProcedureSQL1 = "CREATE PROCEDURE " + procName1 + "\n" +
+                    "    @big_decimal_type      decimal(15, 3),\n" +
+                    "    @big_decimal_type_o    decimal(15, 3) OUTPUT\n" +
+                    "AS\n" +
```
Comment (lilgreenbird) [inline src/test/java/com/microsoft/sqlserver/jdbc/unit/statement/BigDecimalPrecisionTest.java:38]: add boundary cases for max scale 38. see BigDecimal tests in StatementTest.java

**Review by divang — APPROVED** [review]

**Review by tkyc — APPROVED** [review]

**Comment by Ananya2** [comment]
Changes have already been requested as a part of https://github.com/microsoft/mssql-jdbc/pull/2582, hence closing the PR.
